src/Npgsql/VolatileResourceManager.cs

```diff
@@ -292,13 +332,22 @@ void RollbackTwoPhase()
 
         #region Dispose/Cleanup
 
-        void Dispose()
+        public void Dispose()
         {
             if (_isDisposed)
                 return;
             Debug.Assert(_transaction != null, "No transaction");
             Debug.Assert(_connector != null, "No connector");
 
+            // Avoid un-enlisting another RM in case the connection has been concurrently re-used in a new
+            // transaction.
+            if (_connector.EnlistedResource == this)
+                _connector.EnlistedResource = null;
```

Comment:
In the case a prepare phase has been run, we have already set it to null, so with the test, no risk, even with concurrency, to override another RM.
In the rollback case, concurrency is still locked at that point.